File: sem4/mpp/project-cs/repository/ChildDBRepository.cs
using project_cs.domain;
using log4net;


namespace project_cs.repo;

public class ChildDbRepository : IChildRepository {
    private static readonly ILog Log =
        LogManager.GetLogger(System.Reflection.MethodBase.GetCurrentMethod().DeclaringType);

    readonly IDictionary<String, string> _props;

    public ChildDbRepository(IDictionary<string, string> props) {
        Log.Info("creating child repo ");
        this._props = props;
    }

    public Child Add(Child elem) {
        Log.Info("add child");
        var con = DbUtils.GetConnection(_props);
        using (var comm = con.CreateCommand()) {
            comm.CommandText = "INSERT INTO Child (name, age) VALUES(@name, @age)";
            var paramName = comm.CreateParameter();
            paramName.ParameterName = "@name";
            paramName.Value = elem.name;
            comm.Parameters.Add(paramName);

            var paramAge = comm.CreateParameter();
            paramAge.ParameterName = "@age";
            paramAge.Value = elem.age;
            comm.Parameters.Add(paramAge);
            comm.ExecuteNonQuery();

            var comm2 = con.CreateCommand();
            comm2.CommandText = "SELECT last_insert_rowid()";
            var id = (Int64)comm2.ExecuteScalar()!;
            elem.id = (int)id;
            return elem;
        }
    }

    public void Delete(int id) {
        Log.Info("delete child");
        var con = DbUtils.GetConnection(_props);
        using (var comm = con.CreateCommand()) {
            comm.CommandText = "DELETE FROM Child WHERE ID=@id";
            var paramId = comm.CreateParameter();
            paramId.ParameterName = "@id";
            paramId.Value = id;
            comm.Parameters.Add(paramId);

            comm.ExecuteNonQuery();
        }
    }

    public void Update(Child elem, int id) {
        Log.Info("update child");
        var con = DbUtils.GetConnection(_props);
        using (var comm = con.CreateCommand()) {
            comm.CommandText = "UPDATE Child SET name=@name, age=@age WHERE ID=@id";
            var paramId = comm.CreateParameter();
            paramId.ParameterName = "@id";
            paramId.Value = id;
            comm.Parameters.Add(paramId);

            var paramName = comm.CreateParameter();
            paramName.ParameterName = "@name";
            paramName.Value = elem.name;
            comm.Parameters.Add(paramName);

            var paramAge = comm.CreateParameter();
            paramAge.ParameterName = "@age";
            paramAge.Value = elem.age;
            comm.Parameters.Add(paramAge);

            comm.ExecuteNonQuery();
        }
    }

    public Child? GetById(int id) {
        Log.Info("get child");
        var con = DbUtils.GetConnection(_props);
        Child? c = null;
        using (var comm = con.CreateCommand()) {
            comm.CommandText = "SELECT name, age, id FROM child WHERE id=@id";
            var paramId = comm.CreateParameter();
            paramId.ParameterName = "@id";
            paramId.Value = id;
            comm.Parameters.Add(paramId);
            using (var dataR = comm.ExecuteReader()) {
                if (dataR.Read()) {
                    var name = dataR.GetString(0);
                    var age = dataR.GetInt32(1);
                    c = new Child(name, age);
                    c.id = id;
                }
            }
        }

        return c;
    }

    public ICollection<Child> GetAll() {
        Log.Info("get all from child");
        var con = DbUtils.GetConnection(_props);
        IList<Child> children = new List<Child>();
        using (var comm = con.CreateCommand()) {
            comm.CommandText = "SELECT name, age, id FROM child";

            using (var dataR = comm.ExecuteReader()) {
                while (dataR.Read()) {
                    var name = dataR.GetString(0);
                    var age = dataR.GetInt32(1);
                    var id = dataR.GetInt32(2);
                    var c = new Child(name, age);
                    c.id = id;
                    children.Add(c);
                }
            }
        }

        return children;
    }

    public Child? GetChildByNameAge(string name, int age) {
        Log.Info("get child");
        var con = DbUtils.GetConnection(_props);
        Child? c = null;
        using (var comm = con.CreateCommand()) {
            comm.CommandText = "SELECT name, age, id FROM child WHERE name=@name AND age=@age";
            var paramName = comm.CreateParameter();
            paramName.ParameterName = "@name";
            paramName.Value = name;
            comm.Parameters.Add(paramName);
            var paramAge = comm.CreateParameter();
            paramAge.ParameterName = "@age";
            paramAge.Value = age;
            comm.Parameters.Add(paramAge);
            using (var dataR = comm.ExecuteReader()) {
                if (dataR.Read()) {
                    c = new Child(name, age);
                    c.id = dataR.GetInt32(2);
                }
            }
        }

        return c;
    }
}

File: sem4/mpp/project-java/repository/src/main/java/repository/ChildDBRepository.java
package repository;

import domain.Child;
import utils.JdbcUtils;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.sql.*;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.Properties;

public class ChildDBRepository implements ChildRepository {
	private final JdbcUtils dbUtils;
	private static final Logger logger = LogManager.getLogger(ChildDBRepository.class.getName());

	public ChildDBRepository(Properties props) {
		logger.info("Initializing child repository with properties: {} ", props);
		dbUtils = new JdbcUtils(props);
	}

	@Override
	public Child add(Child elem) {
		logger.traceEntry("add child {}", elem);
		Connection con = dbUtils.getConnection();
		try (PreparedStatement preparedStatement = con.prepareStatement("INSERT INTO Child (name, age) values (?,?)")) {
			preparedStatement.setString(1, elem.getName());
			preparedStatement.setInt(2, elem.getAge());
			int result = preparedStatement.executeUpdate();
			logger.trace("saved {}", result);
			ResultSet resultSet = con.prepareStatement("SELECT last_insert_rowid()").executeQuery();
			resultSet.next();
			elem.setId(resultSet.getInt(1));
			logger.trace("set id {}", elem.getId());
		} catch (SQLException e) {
			logger.error(e);
		}
		logger.traceExit();
		return elem;
	}

	@Override
	public void delete(Integer id) {
		logger.traceEntry("delete child");
		Connection con = dbUtils.getConnection();
		try (PreparedStatement preparedStatement = con.prepareStatement("DELETE FROM Child WHERE id=?")) {
			preparedStatement.setInt(1, id);
			int result = preparedStatement.executeUpdate();
			logger.trace("deleted {}", result);
		} catch (SQLException e) {
			logger.error(e);
		}
		logger.traceExit();
	}

	@Override
	public void update(Child elem, Integer id) {
		logger.traceEntry("update child {}", elem);
		Connection con = dbUtils.getConnection();
		try (PreparedStatement preparedStatement = con.prepareStatement("UPDATE Child SET name=?, age=? WHERE id=?")) {
			preparedStatement.setString(1, elem.getName());
			preparedStatement.setInt(2, elem.getAge());
			preparedStatement.setInt(3, id);
			int result = preparedStatement.executeUpdate();
			logger.trace("saved {}", result);
		} catch (SQLException e) {
			logger.error(e);
		}
		logger.traceExit();
	}

	@Override
	public Child getById(Integer id) {
		logger.traceEntry("get children {}");
		Connection con = dbUtils.getConnection();
		Child child = null;
		try (PreparedStatement preparedStatement = con.prepareStatement("SELECT * FROM Child WHERE id=?")) {
			preparedStatement.setInt(1, id);
			try (ResultSet resultset = preparedStatement.executeQuery()) {
				if (resultset.next()) {
					String name = resultset.getString("name");
					int age = resultset.getInt("age");
					child = new Child(name, age);
					child.setId(id);
				} else {
					logger.error("no child with id");
				}
			}
		} catch (SQLException e) {
			logger.error(e);
		}
		logger.traceExit();
		return child;
	}

	@Override
	public Collection<Child> getAll() {
		logger.traceEntry("get children {}");
		Connection con = dbUtils.getConnection();
		List<Child> children = new ArrayList<>();
		try (PreparedStatement preparedStatement = con.prepareStatement("SELECT * FROM Child")) {
			try (ResultSet resultset = preparedStatement.executeQuery()) {
				while (resultset.next()) {
					int id = resultset.getInt("id");
					String name = resultset.getString("name");
					int age = resultset.getInt("age");
					Child child = new Child(name, age);
					child.setId(id);
					children.add(child);
				}
			}
		} catch (SQLException e) {
			logger.error(e);
		}
		logger.traceExit();
		return children;
	}

	@Override
	public Child getChildByNameAge(String name, int age) {
		logger.traceEntry("get if child exists {}");
		Connection con = dbUtils.getConnection();
		Child child = null;
		try (PreparedStatement preparedStatement = con.prepareStatement("SELECT * FROM Child WHERE name=? AND age=?")) {
			preparedStatement.setString(1, name);
			preparedStatement.setInt(2, age);
			try (ResultSet resultset = preparedStatement.executeQuery()) {
				if (resultset.next()) {
					child = new Child(name, age);
					child.setId(resultset.getInt("id"));
				}
			}
		} catch (SQLException e) {
			logger.error(e);
		}
		logger.traceExit();
		return child;
	}
}
